### PR TITLE
[WIP] Api for creating cloud volume snapshot

### PR DESCRIPTION
**Pull request by jeffibm**
Api for creating cloud volume snapshot

UI PR - https://github.com/ManageIQ/manageiq-ui-classic/pull/8245

@miq-bot add-reviewer @kbrock 
@miq-bot add-label enhancement
@miq-bot assign @kbrock

### CODE REVIEW

**Comment by kavyanekkalapu** [comment]
I thought @kbrock is already added snapshotsapi in https://github.com/ManageIQ/manageiq-api/pull/1134. What is this pr do?

File: app/controllers/api/cloud_volumes_controller.rb [hunk 1]
```diff
@@ -42,7 +42,15 @@ def create_backup_resource(type, id, data)
...
-        {:task_id => cloud_volume.backup_restore_queue(User.current_userid, backup['ems_ref'])}
+        {:task_id => cloud_volume.create_volume_snapshot_queue(User.current_userid, backup['ems_ref'])}
+      end
+    end
+
+    def create_snapshot_resource(type, id, data)
+      options = {}
+      options[:name] = data['resources']['snapshot_name'] if data['resources']['snapshot_name']
+      api_resource(type, id, "Restoring backup for", :supports => :snapshot_create) do |cloud_volume|
+        {:task_id => cloud_volume.backup_restore_queue(User.current_userid, options)}
```
Comment (kbrock) [inline app/controllers/api/cloud_volumes_controller.rb:53]: @jeffibm did these two `cloud_volume` methods get swapped by mistake?

File: spec/requests/cloud_volumes_spec.rb [hunk 2]
```diff
@@ -310,4 +310,20 @@
...
+      provider = FactoryBot.create(:ems_autosde, :zone => zone)
+      cloud_volume = FactoryBot.create(:cloud_volume_autosde, :ext_management_system => provider)
+
+      api_basic_authorize(action_identifier(:cloud_volumes, :create_snapshot, :resource_actions, :post))
+      stub_supports(cloud_volume.class, :snapshot_create)
+
+      payload = {:action => "create_snapshot", :resources => {:snapshot_name => "stub_snapshot_name"}}
+      post(api_cloud_volume_url(nil, cloud_volume), :params => payload)
+
+      expect(response).to have_http_status(:ok)
```
Comment (kbrock) [inline spec/requests/cloud_volumes_spec.rb:326]: If it works, please use `expect_single_action_result` with something like `:success => true, :message => /Creating/, :task => true`. You can probably find a lot of examples of this method.

Just saying "it didn't return an error" will catch fewer errors than expecting a certain message back.

File: config/api.yml [hunk 3]
```diff
@@ -814,6 +814,8 @@
         :identifier: cloud_volume_safe_delete
       - :name: create_backup
         :identifier: cloud_volume_backup_create
+      - :name: create_snapshot
+        :identifier: cloud_volume_snapshot_create
```
Comment (kbrock) [inline config/api.yml:818]: glad this already exists.

**Review by kbrock — COMMENTED** [review]
the code looks good, just a little confused about the payload going into `api_resource`.

File: app/controllers/api/cloud_volumes_controller.rb (hunk 1 above)
Comment (jeffibm) [reply app/controllers/api/cloud_volumes_controller.rb]: yes, I had few more changes to be pushed..

**Comment by jeffibm** [comment]
> I thought @kbrock is already added snapshots api in #1134. What is this pr do?

Earlier it was being handled in 2 different controllers. Let me try to combine/reuse them..

**Comment by jeffibm** [comment]
This was handled in https://github.com/ManageIQ/manageiq-api/pull/1162. Hence closing the PR.

**Pull request closed without merging**